WIP: "puzzle show" command

dandrake opened the pull request:
It seems the `puzzle show` command works, and that my refactoring has not affected the `gen` or `solve` commands. Please review. When this looks good, I think I can remove the TODO file -- though I still would like to somehow get the HTML output to include the puzzle code. 

(For that matter, I think the terminal / text output should have that too? then we could eliminate that from the generate and show commands and just have each puzzle, when emitted to whatever format, include its code so the reader can show or solve it later.)

This PR will close #15.

dandrake committed:
initial working version of a 'puzzle show' command

dandrake committed:
add TODO.md with, uh, things to do

dandrake committed:
rename TODO.md to TODO.org

Okay, I'll just admit that I'm too much of an emacs guy and just prefer
org for these kind of notes over markdown...

dandrake committed:
plan for refactoring...

dandrake committed:
initial refactoring for gen, show, and solve: processing input

The generate, show, and solve commands now delete processing the input
and getting the data they need for the rest of their work to two new
functions and get data from a puzzle type, or a puzzle code.

Added a new CommandData record to bundle those pieces of data together.

The initial part of all three command handlers is now delightfully
short: one line to validate the input, then one line to process it into
the data it needs to do the rest of its job.

dandrake committed:
generate and show commands use common method for output

Their error messages differ -- bad code, or bad type -- but once we have
a puzzle to display, both commands do exactly the same thing.

I could likewise factor out a method for the solve command, but since
that code is only ever used by the solve command, it doesn't seem to
have any benefit.

dandrake committed:
fix tests broken by changed 'help' output

There are still tests that fail on my Ubuntu machine, related to fonts
(see #17), and when I first ran the tests, they gobbled up all my memory
(I have 32GB on this computer) and made the entire thing unresponsive
for a while. Likely not something we need to address, but FYI.

pcantrell commented on test/fixtures/integration/help.expected.log:
```diff
@@ -4,10 +4,12 @@ Picked up _JAVA_OPTIONS: -Dfile.encoding=UTF-8
 
 Options:
   -d,--difficulty <num>      Change puzzle difficulty from default
+                             Use -1 to see range and minimum for credit
```
Oh, that's a good idea! Let's make it `-d help` or similar; special-casing -1 feels just a little clumsy.

pcantrell commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -226,6 +260,26 @@ private static void applyOptionsToContext(
         }
     }
 
+    private static void displayPuzzle(Puzzle puzzle, PuzzleContext ctx, PuzzleOptions options) throws IOException {
```
This should just take a single `CommandData` param!

pcantrell commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -144,33 +157,57 @@ private static void solve(PuzzleOptions options) throws InvalidPuzzleCodeExcepti
         Puzzle puzzle,
-        boolean solutionOutput
-    ) throws IOException {
+        boolean solutionOutput) throws IOException {
```
(Project formatting conventions again)

pcantrell commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -226,6 +260,26 @@ private static void applyOptionsToContext(
+            var solveCtx = ctx.cleanCopy();
+            var puzzleForSolution = Puzzle.findByID(puzzle.id());
+            boolean solutionOutput = true;
+            applyOptionsToContext(options, solveCtx, puzzleForSolution, solutionOutput);
```
Yeah, Java really needs parameter labels already!

_Most_ Java IDEs these days show labels for _most_ args in _most_ situations where it's helpful…but it's inconsistent and often harms readability more than it helps. I’ve configured IntelliJ to hide all of its “inlay hints” at first (that's what it calls them) but toggle them with a key shortcut when I wonder what a param is for.

Long story short, I see the original code as `solutionOutput: false` in IntelliJ, but I like your solution better: it works for everyone all the time.

pcantrell commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -80,59 +85,67 @@ private static void requireCommandArgs(int expectedArgCount, PuzzleOptions optio
-            applyOptionsToContext(options, solveCtx, puzzleForSolution, true);
-            emitPuzzle(puzzleForSolution, solveCtx, options);
+        // output the result:
+        if (data.puzzle == null) {
```
Would it make more sense to have `getPuzzleFromCode` return null for `data` in its entirety, not just `data.puzzle`? Feels cleaner to me…maybe.

dandrake replied:
I could go either way. I do like the specificity of looking at `puzzle`. 

But: can we actually *get* there? If you pass in a bad code, will `getPuzzleFromCode` always throw an `InvalidPuzzleCodeException`, which will bubble up and not even execute the check for a null puzzle?

Do you have any examples of codes that won't trigger the exception, but for which it won't find a puzzle?

pcantrell commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -144,33 +157,57 @@ private static void solve(PuzzleOptions options) throws InvalidPuzzleCodeExcepti
         }
     }
 
+    private static CommandData getPuzzleFromType(PuzzleOptions options) throws IOException {
```
Suggested rename: `generatePuzzleFromType`

pcantrell reviewed (commented):
Looks good! I left a couple of substantive ideas / questions about further improvement, plus bunch of fiddly things (mostly pleas about formatting). Notwithstanding all that, the big picture is nice feature, solid approach.

dandrake committed:
change CommandData -> PuzzleOutputContext

renamed based on PR comment

dandrake committed:
help text update for 'puzzle show'

dandrake commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -49,12 +49,17 @@ public static void main(String[] args) {
             }
-        } catch(Exception e) {
+        } catch (InvalidPuzzleCodeException e) {
+            System.out.println("invalid puzzle code exception");
```
The generic `Exception e` below does catch that, but I think it's good practice to be more specific with the exceptions we catch (inspired by Effective Java, etc) and to provide a better error message. Especially since the bits shown the end (the stacktrace, info, and scan stuff under "Tryt:") are gradle-specific and don't actually work.

OTOH if I remove that, it does mention the exception, and one could reasonably expect a student in 127 to work out what happened. 

Thoughts? I can go either way. If we do keep it, I do think I should output to stderr and make the error message cleaner and nicer.

dandrake commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -80,59 +85,67 @@ private static void requireCommandArgs(int expectedArgCount, PuzzleOptions optio
+
+        // output the result:
+        if (data.puzzle == null) {
+            String puzzleType = options.commandAndArgs().get(1);
```
I see two places where it gets a type, and one where it gets a puzzle code. I'm not sure the juice is worth the squeeze here; it's just three places, and two different contexts/purposes.

Are you thinking of generic methods on `options` that look at the command, and use that to use `get(1)`? I'm thinking:

* `options.getPuzzleType()` would look at the command (gen) and output the string if so, and null otherwise;
* `options.getPuzzleCode()` would do the same thing when the command is `show` or `solve`.

That said, the three places where we just use `get(1)` are clear enough. It is true that, if one is being very fussy, we are mixing levels of abstraction (accessing a specific index of a list versus the higher-level semantics of "a puzzle code" or "a puzzle type") which isn't the very best style, but... :shrug:

pcantrell replied:
Oh right, I'd missed the mixed semantics of that `get(1)`. Probably not worth the trouble.

dandrake commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -144,33 +157,57 @@ private static void solve(PuzzleOptions options) throws InvalidPuzzleCodeExcepti
+            ctx = PuzzleContext.generate(puzzle.id(),
+                                         options.difficulty() != null
+                                         ? options.difficulty()
+                                         : puzzle.goalDifficulty());
```
I'm currently seeing the formatting according to the first one above -- and if I fiddle with it, then do control-shift-I (reformat, on Linux) or save, it reformats back to your first suggestion. In my `settings.json` on this machine I have `"editor.formatOnSave": true,`.

At any rate, as far as indentation holy wars:

* I don't have any strong feelings about the indentation and formatting
* I *do* strongly feel that we should have something automated (or nearly so) that does the formatting for us.

For example, with JavaScript and web stuff, I just run everything through `prettier`.

Is the default VS Code reformatting okay here?

dandrake replied:
I think I've cleaned up the spurious differences? See e2d37e1136ae1fc5da24f6f75da02bb1108c996f. I'm looking at the overall diff and I *think* everything there is meaningfully related to my code changes and not from new indentation, line breaks, etc. 

If there's anything else you see, I'll fix that up.

dandrake committed:
displayPuzzle takes single PuzzleOutputContext param, not 3

dandrake committed:
fix spelling typo

dandrake committed:
rename getPuzzleFromType to say "generate"

dandrake commented on src/edu/macalester/conceptual/cli/PuzzleOptions.java:
```diff
@@ -26,7 +26,8 @@ class PuzzleOptions {
         repeat = addOption("r", "repeat", "num", "Generate <num> different puzzles");
-        difficulty = addOption("d", "difficulty", "num", "Change puzzle difficulty from default");
+        difficulty = addOption("d", "difficulty", "num", "Change puzzle difficulty from default\n" +
+            "Use -1 to see range and minimum for credit");
```
This is now #20.

dandrake commented on src/edu/macalester/conceptual/cli/CommandLine.java:
```diff
@@ -144,33 +157,57 @@ private static void solve(PuzzleOptions options) throws InvalidPuzzleCodeExcepti
+    private static CommandData getPuzzleFromType(PuzzleOptions options) throws IOException {
+        // the commandline UI refers to "types"; the Puzzle class and
+        // friends use "name" because they talk about *names* of
+        // classes.
```
this is now #19.

dandrake committed:
remove InvalidPuzzleCodeException handling

See issue #21. We'll figure out what to do here later.

dandrake committed:
clean up some spurious differences from VS Code autoformatting

Trying to match project conventions.

dandrake committed:
fix failed tests related to minor help text change

I changed "...print a puzzle" to "print a previously generated puzzle".
This fixes up the expected output for those tests.
